### PR TITLE
`ultralytics 8.4.8` Support `max_det` and `agnostic_nms` for end2end

### LINKED ISSUE #23337: The max_det parameter of YOLO26 does not work
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

The max_det parameter of YOLO26 does not work，Even if max_det is set to 5000, the maximum number of identifications is still 300

### Environment

(yolo11) PS E:\A_worker\W_XM\Light_train> yolo checks
Ultralytics 8.4.6  Python-3.12.12 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB)
Setup complete  (20 CPUs, 31.7 GB RAM, 857.4/953.9 GB disk)

OS                     Windows-11-10.0.26100-SP0
Environment            Windows
Python                 3.12.12
Install                pip
Path                   E:\Miniconda\envs\yolo11\Lib\site-packages\ultralytics
RAM                    31.71 GB
Disk                   857.4/953.9 GB
CPU                    12th Gen Intel Core i7-12700H
CPU count              20
GPU                    NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB
GPU count              1
CUDA                   12.8

numpy                   2.4.1>=1.23.0
matplotlib              3.10.8>=3.3.0
opencv-python           4.13.0.90>=4.6.0
pillow                  12.0.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.9.1+cu128>=1.8.0
torch                   2.9.1+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.1+cu128>=0.9.0
psutil                  7.1.3>=5.8.0
polars                  1.36.1>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
(yolo11) PS E:\A_worker\W_XM\Light_train> 

### Minimal Reproducible Example

from ultralytics import YOLO
import torch
from tqdm import tqdm
import cv2  # 需要导入cv2来获取视频的总帧数

# 检查是否可以使用GPU
print(torch.cuda.is_available())

# 加载训练好的YOLO模型
model = YOLO(r"E:\A_worker\W_XM\yolo11_car_seg\models\car_type_v21_1920.pt",)
             # task='detect')

# 视频路径
video_path = r"E:\A_worker\W_XM\yolo11_car_seg\test_img\flow\1.mp4"

# 使用OpenCV获取视频的总帧数，用于初始化进度条
cap = cv2.VideoCapture(video_path)
total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
cap.release()  # 立即释放资源

# 初始化进度条
pbar = tqdm(total=total_frames, desc="Processing video", unit="frames")

# 预测视频
results = model.track(
    source=video_path,
    save=True,
    save_txt=False,
    show=True,
    imgsz=1920,
    # iou=0.2,
    classes=[1],
    verbose=False,
    # line_width=2,
    # show_labels=False,
    # conf=0.5,
    tracker=r'E:\A_worker\W_XM\yolo11_car_seg\botsort_customs.yaml',
    # persist=True,
    stream=True  # 使用流式处理以避免内存问题
)

# 处理每一帧并更新进度条
for result in results:
    pbar.update(1)  # 每处理完一帧，进度条更新

# 关闭进度条
pbar.close()

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #23346: YOLO26n导出tflite后，索引为0的类别会出现与上一条类别定位完全重合的错误
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

YOLO26n导出tflite后，索引为0的类别会出现与上一条类别定位完全重合的错误

### Additional

我们简单训练了一个ocr识别数字和字母的demo，在.pt下测试是完全正确的，而在导出tflite后会出现以下问题：
0 = {float[6]@47206} [0.104787946, 0.46824303, 0.18168591, 0.58257985, 0.96063465, 24.0]
**1 = {float[6]@47207} [0.019556295, 0.4436751, 0.107705675, 0.5854393, 0.9603612, 23.0]
2 = {float[6]@47208} [0.019556295, 0.4436751, 0.107705675, 0.5854393, 0.9579195, 0.0]
3 = {float[6]@47209} [0.28735766, 0.4436547, 0.36102203, 0.56042284, 0.95364594, 3.0]
4 = {float[6]@47210} [0.28735766, 0.4436547, 0.36102203, 0.56042284, 0.9529765, 0.0]**
5 = {float[6]@47211} [0.38027313, 0.44484878, 0.4585964, 0.56000495, 0.95124876, 0.0]
6 = {float[6]@47212} [0.47619057, 0.44451082, 0.5499851, 0.5611348, 0.95100635, 8.0]
7 = {float[6]@47213} [0.8529782, 0.44925645, 0.93138236, 0.5644056, 0.9507863, 7.0]
8 = {float[6]@47214} [0.6650287, 0.44675952, 0.7398499, 0.5633069, 0.94486284, 7.0]
9 = {float[6]@47215} [0.1915837, 0.44322687, 0.2707733, 0.5611132, 0.9418906, 2.0]
后面的省略

Image

以下是我处理后的结果：
0 = {Recognition@47194} Recognition(title=O, confidence=0.96063465, location=Rect(76, 338 - 131, 420))
1 = {Recognition@47195} Recognition(title=N, confidence=0.9603612, location=**Rect(14, 320 - 78, 422)**)
2 = {Recognition@47196} Recognition(title=0, confidence=0.9579195, location=**Rect(14, 320 - 78, 422)**)
3 = {Recognition@47197} Recognition(title=3, confidence=0.95364594, location=**Rect(207, 320 - 260, 404)**)
4 = {Recognition@47198} Recognition(title=0, confidence=0.9529765, location=**Rect(207, 320 - 260, 404)**)
5 = {Recognition@47199} Recognition(title=0, confidence=0.95124876, location=Rect(274, 321 - 331, 404))
6 = {Recognition@47200} Recognition(title=8, confidence=0.95100635, location=Rect(343, 320 - 397, 405))
7 = {Recognition@47201} Recognition(title=7, confidence=0.9507863, location=Rect(615, 324 - 672, 407))
8 = {Recognition@47202} Recognition(title=7, confidence=0.94486284, location=Rect(479, 322 - 533, 406))
9 = {Recognition@47203} Recognition(title=2, confidence=0.9418906, location=Rect(138, 320 - 195, 405))

Image

### PR DESCRIPTION
**Pull request by Y-T-G**
Fixes #23337 
Fixes https://github.com/ultralytics/ultralytics/issues/23383
Fixes https://github.com/ultralytics/ultralytics/issues/23346

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds better control over end-to-end (NMS-free) YOLO inference by wiring `max_det` and `agnostic_nms` into model heads, updating docs/tests, and bumping the package version 🚀

### 📊 Key Changes
- Introduced a generic `set_head_attr(**kwargs)` helper on YOLO models to safely set head (last-layer) attributes like `end2end`, `max_det`, and `agnostic_nms` 🧩
- When `end2end=True`, **predict** and **val** now push `max_det` and `agnostic_nms` into the model head so these settings actually take effect ⚙️
- Added `agnostic_nms` support to the `Detect` head (affects top-k selection logic used in end-to-end flows) 🎯
- Export pipeline now passes `agnostic_nms` into the exported head, improving parity between Python inference and exported formats 📦
- Training path sets `max_det` on end-to-end models during attribute setup ✅
- Docs updated:
  - `train` args now document `max_det`
  - `predict/val` docs for `end2end` no longer claim it enables using `agnostic_nms` via the “traditional NMS pipeline” 📚
- CLI tests updated to explicitly include `max_det=100` (and `agnostic_nms` in val) to validate these pathways 🧪
- Version bump: `8.4.7` → `8.4.8` 🔖

### 🎯 Purpose & Impact
- Makes `max_det` and `agnostic_nms` behave consistently for **end-to-end (NMS-free)** models like YOLO26 and YOLOv10, rather than being ignored in some flows 🔧
- Improves reliability and reproducibility across **predict**, **validate**, **train-time validation**, and **exported** deployments (fewer “works in Python but not after export” surprises) 📈
- Clarifies documentation so users better understand what `end2end=False` enables (traditional NMS + `iou`), and ensures `max_det` is visible as a training/validation control 📝

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review—an Ultralytics engineer will assist soon. 🧩

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -235,11 +236,17 @@ def get_topk_index(self, scores: torch.Tensor, max_det: int) -> tuple[torch.Tens
         # Use max_det directly during export for TensorRT compatibility (requires k to be constant),
         # otherwise use min(max_det, anchors) for safety with small inputs during Python inference
         k = max_det if self.export else min(max_det, anchors)
-        ori_index = scores.max(dim=-1)[0].topk(k)[1].unsqueeze(-1)
-        scores = scores.gather(dim=1, index=ori_index.repeat(1, 1, nc))
-        scores, index = scores.flatten(1).topk(k)
-        idx = ori_index[torch.arange(batch_size)[..., None], index // nc]  # original index
-        return scores[..., None], (index % nc)[..., None].float(), idx
+        if self.agnostic_nms:
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/head.py:239]: ⚠️ **HIGH**: In the agnostic branch, `scores.topk(k, dim=1)` runs on a tensor shaped (B, A, 1), so `indices` will be (B, k, 1). Downstream code typically expects an index tensor shaped (B, k). This extra singleton dimension can cause incorrect indexing or shape mismatches when gathering boxes. Consider squeezing the last dimension (or avoid the trailing singleton by squeezing `scores` before topk).

File: ultralytics/nn/modules/head.py [hunk 2]
```diff
@@ -235,11 +236,17 @@ def get_topk_index(self, scores: torch.Tensor, max_det: int) -> tuple[torch.Tens
...
+            scores, labels = scores.max(dim=-1, keepdim=True)
+            scores, indices = scores.topk(k, dim=1)
+            labels = labels.gather(1, indices)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/head.py:242]: 💡 **MEDIUM**: `labels` returned from `scores.max`/`gather` is `long`, whereas the non-agnostic path returns `float()`. If downstream assumes float class IDs (consistent with existing return type), this change can introduce type mismatches. Consider casting `labels = labels.float()` for consistency.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall looks solid and the propagation of `max_det`/`agnostic_nms` is consistent. Only potential issues are in `get_topk_index` agnostic branch: the returned index tensor likely has an extra singleton dim, and label dtype changes from float to long. Addressing those should make the PR clean.

💬 Posted 2 inline comments

**Comment by Y-T-G** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-01-26 09:28:04 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/21352595010): `ci.yml` run 17239
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/21352595716): `docker.yml` run 1745

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-01-27 07:25:35 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/21388328546): `ci.yml` run 17276
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/21388329038): `docker.yml` run 1748

**Comment by UltralyticsAssistant** [comment]
Merged — huge thanks @Y-T-G (and @Laughing-q, @glenn-jocher) for landing this! This is a meaningful step forward for end-to-end (NMS-free) inference consistency across Ultralytics.

> “If I have seen further it is by standing on the shoulders of giants.” — Isaac Newton

This PR really embodies that: by wiring `max_det` and `agnostic_nms` directly into the model head (plus the new `set_head_attr(**kwargs)` helper), you’ve tightened parity across **predict**, **val**, **train-time validation**, and **export**—exactly the kind of reliability that makes deploying models like **YOLO26** feel predictable and reproducible. The doc/test updates and version bump are the polish that turns a good fix into a great user experience.

Appreciate the careful engineering and teamwork here—thank you all for making Ultralytics better.
